html/captive-portal/lib/captiveportal/PacketFence/Controller/Pay.pm

```diff
@@ -256,40 +190,382 @@ sub showBilling : Private {
+        mac         => $mac,
+        item        => $tier,
+        price       => $tiers_infos{$tier}{'price'},
+        description => $tiers_infos{$tier}{'description'},
+    };
+    $c->stash(
+        template    => 'billing/mirapay_iframe_pay.html',
+        mirapay_url => $self->_build_mirapay_url($c, $transaction_infos_ref)
+    );
+}
+
+=head2 _build_mirapay_url
+
+=cut
+
+sub _build_mirapay_url {
+    my ($self, $c, $transaction_infos_ref) = @_;
+    my $request      = $c->request;
+    my $billing      = $Config{billing};
+    my $url          = $billing->{mirapay_iframe_url};
+    my $merchant_id  = $billing->{mirapay_iframe_merchant_id};
+    my $cardholdername = $request->param('firstname') . ' ' . $request->param('lastname') ;
+    my $redirect_url = "https://" . join(".",$Config{general}{hostname},$Config{general}{domain}) . "/pay/mirapay_iframe_process";
+    my @params       = (
+        MerchantID  => $merchant_id,
+        RedirectURL => $redirect_url,
+        EchoData    => $transaction_infos_ref->{item},
+        Amount      => $transaction_infos_ref->{price} * 100
+    );
+    my $mkey = $self->calc_mkey($c, @params);
+    my $query = join("&", pairmap {"$a=" . uri_escape($b)} @params, 'MKEY', $mkey);
+    return "$url?$query";
+}
+
+=head2 mirapay_iframe_process
+
+=cut
+
+sub mirapay_iframe_process : Local {
+    my ($self, $c) = @_;
+    if ($self->_mirapay_verify_request($c)) {
+        $c->stash->{template} = 'billing/mirapay_iframe_process.html';
+    }
+    else {
+        $c->stash->{template} = 'billing/mirapay_iframe_process_error.html';
+    }
+}
+
+=head2 mirapay_iframe_process_success
+
+=cut
+
+sub mirapay_iframe_process_success : Local {
+    my ($self, $c) = @_;
+    my $billingObj    = new pf::billing::custom();
+    my $request       = $c->request;
+    my $session       = $c->session;
+    my $logger        = $c->log;
+    my $portalSession = $c->portalSession;
+    my $profile       = $c->profile;
+    my $mac           = $portalSession->clientMac;
+    my $approval_code = $request->param('ApprovalCode');
+    my $bank_response = $request->param('BankResponse');
+    my $cvv_response  = $request->param('CVVResponse');
+    my $date_time     = $request->param('DateTime');
+    my $avs_response  = $request->param('AVSResponse');
+
+    # Transactions informations
+    my $tier                  = $session->{'tier'};
+    my %tiers_infos           = $billingObj->getAvailableTiers();
+    my $transaction_infos_ref = {
+        ip          => $portalSession->clientIp(),
+        mac         => $mac,
+        firstname   => $session->{'firstname'},
+        lastname    => $session->{'lastname'},
+        email       => lc($session->{'email'}),
+        item        => $tier,
+        price       => $tiers_infos{$tier}{'price'},
+        description => $tiers_infos{$tier}{'description'},
+    };
+    $c->forward('processSuccessPayment', [$tier, $transaction_infos_ref]);
+}
+
+our %MirapayResponseCode = (
```

Comment:
In the middle of the file ?
Also that should be read only and in a constants file